### DISCUSSION TITLE
Forward declare let?

### DISCUSSION THREAD

**Discussion by Timmmm** [rems-project/sail#795] [Q&A] [answered]
Is it possible to forward-declare a global `let` constant? Something like this?

```
val default_sv32_ext_pte : pte_ext_bits

...

let default_sv32_ext_pte : pte_ext_bits = zeros()
```

Easy to work around by changing it to a function; just wondering if there was a better way.

**Comment by Alasdair** [comment, marked as answer]
We don't currently allow this. I think circular let-bindings would cause issues for some of our backends currently as they assume all the global let-bound constants can just be initialised in order (and some cases like below would just be impossible to handle). Also we might have issues with dependency cycles involving functions and forward-declared constants.
```
val x : int
val y : int

let x = 1 + y
let y = 1 + x
```

**Reply by Alasdair** [reply]
Not to say it'd be impossible we'd just have to have some additional check to ensure these cases couldn't occur.
